File: practice/data_structures/linked_lists/insert_node_into_a_sorted_doubly_linked_list/insert_node_into_a_sorted_doubly_linked_list.py
"""
 Insert a node into a sorted doubly linked list
 head could be None as well for empty list
 Node is defined as
 
 class Node(object):
 
   def __init__(self, data=None, next_node=None, prev_node = None):
       self.data = data
       self.next = next_node
       self.prev = prev_node

 return the head node of the updated list 
"""

def SortedInsert(i_head, i_data):
    if not i_head:
        return Node(i_data)
    if i_head.data > i_data:
        n = Node(i_data, i_head)
        i_head.prev = n
        i_head = n
        return i_head
    
    cp = i_head
    while cp.next and cp.data < i_data:
        cp = cp.next
    if not cp.next and cp.data < i_data:
        n = Node(i_data, None, cp)
        cp.next = n
    else:
        n = Node(i_data, cp, cp.prev)
        cp.prev.next = n
        cp.prev = n
    return i_head

File: practice/data_structures/linked_lists/insert_node_into_a_sorted_doubly_linked_list/insert_node_into_a_sorted_doubly_linked_list.cpp
/*
    Insert Node in a doubly sorted linked list 
    After each insertion, the list should be sorted
   Node is defined as
   struct Node
   {
     int data;
     Node *next;
     Node *prev;
   }
*/

Node * SortedInsert(Node * i_head, int i_data)
{
    if (!i_head)
    {
        Node * head = new Node;
        head->prev = head->next = nullptr;
        head->data = i_data;
        return head;
    }
    
    if (i_head->data > i_data)
    {
        Node * head = new Node;
        head->prev = nullptr;
        head->next = i_head;
        head->data = i_data;
        
        i_head->prev = head;
        return head;
    }
    
    Node * node = i_head;
    while (node->next && (node->data < i_data))
    {
        node = node->next;
    }
    if (!node->next && (node->data < i_data))
    {
        Node * last = new Node;
        last->data = i_data;
        last->next = nullptr;
        last->prev = node;
        node->next = last;
    }
    else
    {
        Node * last = new Node;
        last->data = i_data;
        last->next = node;
        last->prev = node->prev;
        
        last->prev->next = last;
        node->prev = last;
    }
    
    return i_head;
}

